packages/time-oracle/src/alarms.rs

```diff
@@ -74,6 +78,49 @@ impl<'a> Alarms<'a> {
     }
 }
 
+pub struct InDelivery<'r> {
+    storage: &'r mut dyn Storage,
+}
+
+impl<'r> InDelivery<'r> {
+    const ALARMS: Deque<'_, TimeAlarm> = Deque::new("in_delivery");
+
+    pub fn new(storage: &'r mut dyn Storage) -> Self {
+        Self { storage }
+    }
+
+    pub fn add(&mut self, subscriber: Addr, time: TimeSeconds) -> Result<(), AlarmError> {
+        Self::ALARMS
+            .push_back(self.storage, &TimeAlarm { subscriber, time })
+            .map_err(Into::into)
+    }
+
+    pub fn delivered(&mut self) -> Result<(), AlarmError> {
+        Self::ALARMS
+            .pop_front(self.storage)
+            .map(|maybe_alarm: Option<TimeAlarm>| debug_assert!(maybe_alarm.is_some()))
+            .map_err(Into::into)
+    }
+
+    pub fn failed(&mut self, alarms: &mut Alarms<'_>) -> Result<(), AlarmError> {
+        Self::ALARMS
+            .pop_front(self.storage)
+            .map_err(Into::into)
+            .and_then(|maybe_alarm: Option<TimeAlarm>| {
+                maybe_alarm.ok_or(AlarmError::ReplyOnEmptyAlarmQueue)
+            })
+            .and_then(|TimeAlarm { subscriber, time }| {
+                alarms.add(self.storage, subscriber, Timestamp::from_seconds(time))
```

Comment:
use a private method of `Alarms::add_impl` to avoid the 'seconds - millis - seconds' conversion